src/Lut.ts

```diff
@@ -0,0 +1,500 @@
+import { getColorByChannelIndex } from "./constants/colors.js";
+import Histogram from "./Histogram.js";
+
+function clamp(val: number, cmin: number, cmax: number): number {
+  return Math.min(Math.max(cmin, val), cmax);
+}
+
+function lerp(xmin, xmax, a) {
+  return a * (xmax - xmin) + xmin;
+}
+
+// We have an intensity value that is in the range of valueMin to valueMax.
+// This domain is assumed to have been remapped from oldMin to oldMax.
+// We now wish to find the intensity value that corresponds to the same relative position in the new domain of newMin to newMax.
+// For our Luts valueMin will always be 0, and valueMax will always be 255.
+// oldMin and oldMax will be the domain of the original raw data intensities.
+// newMin and newMax will be the domain of the new raw data intensities.
+function remapDomain(
+  value: number,
+  valueMin: number,
+  valueMax: number,
+  oldMin: number,
+  oldMax: number,
+  newMin: number,
+  newMax: number
+): number {
+  const pctOfRange = (value - valueMin) / (valueMax - valueMin);
+  const newValue = (newMax - newMin) * pctOfRange + newMin;
+  // now locate this value as a relative index in the old range
+  const pctOfOldRange = (newValue - oldMin) / (oldMax - oldMin);
+  const remapped = valueMin + pctOfOldRange * (valueMax - valueMin);
+  return remapped;
+}
+
+export const LUT_ENTRIES = 256;
+export const LUT_ARRAY_LENGTH = LUT_ENTRIES * 4;
+
+/**
+ * @typedef {Object} ControlPoint Used for the TF (transfer function) editor GUI.
+ * Need to be converted to LUT for rendering.
+ * @property {number} x The X Coordinate
+ * @property {number} opacity The Opacity, from 0 to 1
+ * @property {Array.<number>} color The Color, 3 numbers from 0-255 for r,g,b
+ */
+export type ControlPoint = {
+  x: number;
+  opacity: number;
+  color: [number, number, number];
+};
+
+function controlPointToRGBA(controlPoint) {
+  return [controlPoint.color[0], controlPoint.color[1], controlPoint.color[2], Math.floor(controlPoint.opacity * 255)];
+}
+
+/**
+ * @typedef {Object} Lut Used for rendering.
+ * @property {Array.<number>} lut LUT_ARRAY_LENGTH element lookup table as array
+ * (maps scalar intensity to a rgb color plus alpha, with each value from 0-255)
+ * @property {Array.<ControlPoint>} controlPoints
+ */
+export class Lut {
+  public lut: Uint8Array;
+  public controlPoints: ControlPoint[];
```

Comment:
Having two representations of basically the same information in here makes me a bit nervous, but it seems to be responsibly enough managed in the body of the class and does have enough benefit to make sense.

Reply:
I would think of the `lut` as a cached version of the data array that the `controlPoints` generate.  It's easy to go from controlPoints to lut, but hard to go the other way.  However, control points to lut is not a trivial operation so I keep the lut array around.  We can think about the implications of removing that...  Agave stores even higher level representations like pctMin/pctMax, window/level, isovalue, in addition to control points, based on how the user configured the lut.